sdk/dotnet/Pulumi/Deployment/Deployment_Prepare.cs

```diff
@@ -151,7 +151,12 @@ private static void AddTransitivelyReferencedChildResourcesOfComponentResources(
                 {
                     if (resource is ComponentResource)
                     {
-                        AddTransitivelyReferencedChildResourcesOfComponentResources(resource.ChildResources, result);
+                        HashSet<Resource> childResources = new HashSet<Resource>();
+                        lock (resource.ChildResources)
+                        {
+                            childResources.AddRange(resource.ChildResources);
+                        }
+                        AddTransitivelyReferencedChildResourcesOfComponentResources(childResources, result);
```

Comment:
I doubt it makes much of a measurable difference, but leveraging [`HashSet<T>`'s constructor](https://github.com/dotnet/runtime/blob/477363b27c658bc4f89bab3995386760e95ee8ef/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L86) to do the copy should be more efficient than using our `AddRange` helper. (Haven't actually tried it):

```suggestion
                        HashSet<Resource> childResources;
                        lock (resource.ChildResources)
                        {
                            childResources = new HashSet<Resource>(resource.ChildResources);
                        }
                        AddTransitivelyReferencedChildResourcesOfComponentResources(childResources, result);
```